src/Service/GetMessageRetriesCountService.php

```diff
@@ -0,0 +1,28 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace RunAsRoot\MessageQueueRetry\Service;
+
+use Magento\Framework\MessageQueue\EnvelopeInterface;
+use PhpAmqpLib\Wire\AMQPTable;
+
+class GetMessageRetriesCountService
+{
+    public function execute(EnvelopeInterface $message): int
+    {
+        $messageProperties = $message->getProperties();
+        $applicationHeaders = $messageProperties['application_headers'] ?? null;
+
+        // If there are no application headers, then it is the first time the message has been processed.
+        if (!$applicationHeaders instanceof AMQPTable) {
+            return 0;
+        }
+
+        if (isset($applicationHeaders->getNativeData()['x-death'][0]['count'])) {
+            return $applicationHeaders->getNativeData()['x-death'][0]['count'];
```

Comment:
I removed adding `+1` to the retries count. For me, the first consuming of the message sounds like not a retry but a first try.
So when the message execution failed the first time - the retry count should be 0.
Then it goes to `*_delay` queue, then goes back to the main queue, then it processes, then it fails again - the retry count now is 1. For me, it sounds logical. Here is a step-by-step explanation:

1. message is added to the queue
2. consumer picked it up, but processing failed. try count - 1, retry count - 0
3. message goes to *_delay queue, then goes back to the queue
4. consumer picked it up, but processing failed. try count - 2, retry count - 1
....

According to that if we configure the allowed retries count as 3, it will fail the first time and then will try to process it additional 3 times as configured.

What do you think @mwr @cristiano-pacheco ?